### ISSUE TITLE
Initial stumbling blocks

### ISSUE THREAD

**Issue by kxygk** [lilactown/flex#4]
Hi, 

Sorry I'm getting back to you so late. I appreciate the thoughts you shared with me here:

https://old.reddit.com/r/Clojure/comments/11723ea/is_there_a_reframecljfxlike/

I just tried to convert a little plotting program of mine to use flex and I've hit a few initial stumbling blocks. I only really mean this as to share an initial-impression/user-experience - and not criticism. It's very likely "i'm holding it wrong" and misunderstanding how to use flex.

1. Signals can't be inspected at the REPL

Just working off the example in the README. You hook up `counter-sq` to your `counter` atom. When `counter` updates, `counter-sq` updates under the hood. However as far as I can tell, there is no way to actually inspect `counter-sq` at the REPL.  It just returns a `:town.lilac.flex/init`. You can hook up a listener and print the value, but it's clunky and only prints a single time on change. tbh, I expected it to outwardly continue to behave transparently as a integer - but at the very least I'd want to be able to peak at its value at the REPL :)

2. Not super clear how to chain signals
I obviously wants to go a step further from the basic example and chain my signals. Leverage diamond dependencies and so on.. Since I can't dereference them, I'm not super sure how to do that? I can listen to a signal and create a new signal - but that seems a bit clunky. I don't think it will do the diamond dependency magic that way as well?

3. signal wrapping everything..
Basically any variable that derefs a `flex/source` will in effect be a signal. So I'm envisioning you end up with a ton of this in your code:

```
(def my-var
    (flex/signal
    ....
...
```

Maybe just aesthetic but it'd be a bit clearer with a macro
```
(defs my-var
    ... 
```
Because you don't exactly have a static variable nor really a `defn` function. That said, I get the desire to avoid macros :))

4. No quick and easy way to subscribe to map values
In the CLJFX model the equivalent of signals are split into `sub-val` and `sub-ctx`. The latter is a memoized function call - a bit similar to your signal. The former is a convenience function that allows you to subscribe to map values

So as I started to massage my code to use flex, my `flex/source` atom looked something like this:

```
(def
  params
  "The state atom"
  (flex/source
    {:input-dir   "./data/gauge/"
     :id-2-name   {1645240 5
                   8373316 3
                   7598252 4
                   1683136 2
                   8371400 1
                   1641660 6
                   7598984 7}
     :inside-d18o  "irregular-cave-samples.csv"
     :outside-d18o "manual-rain-samples.csv"
     :plot-height 400
     :plot-width 1000
     :blah        "blah"}))
```

From the API it looks like in flex I need to create a signal for each an every key @_@ so that when a part of the map gets updated I get the appropriate update (that can propagate to all the signals).  Otherwise my whole DAG gets recomputed on each map update (well I assume it'll terminate when identical values are returned). Ergonomically this is a bit of a mess. Basically if you have a .. simple? (finite/non-lazy) built in  datatype as a source, you can be sure you'll automatically want signals for every single elements

in CLJFX `sub-val` is limited to first level map keys - but ideally then should be able to generate a signal on deeper values

Anyway, those are my first thoughts :))

**Comment by lilactown** [comment]
Sure, these are great questions. I'll answer them in order

## 1. Signals can't be inspected at the REPL

Signals can definitely be inspected at the REPL. Here's an example

```clojure
(def counter (flex/source 0))
(def counter-sq (flex/signal (* @counter @counter)))

@counter-sq ;; => :town.lilac.flex/init

(def fx (flex/effect [] @counter-sq))

@counter-sq ;; => 0

(counter 2)

@counter-sq ;; => 4
```

You said

>  It just returns a `:town.lilac.flex/init`. You can hook up a listener and print the value, but it's clunky and only prints a single time on change

What it sounds like is happening is you're dereferencing a signal that hasn't been "connected." By default, signals do not recompute on change. Only if they are depended on by another connected signal or an effect do they "connect" do they compute their value and react to changes. This allows you to construct as many signals as you want, and only those that are referenced will react to updates and store their value, which is nice for letting letting the GC clean things up so we don't take up too much memory.

Ultimately, the goal of flex is to trigger side effects based on changes to inputs. If you only want to inspect a value, you can do what I did in the example above and create a dummy effect that dereferences the signal. You can then dispose of the effect when you're done. But I'm curious, what do you want to happen as these values change?

## 2. Not super clear how to chain signals

> Since I can't dereference them, I'm not super sure how to do that? I can listen to a signal and create a new signal - but that seems a bit clunky. I don't think it will do the diamond dependency magic that way as well?

I'm not sure what you mean here. "Chain my signals" means you want signals that depend on other signals, right? That is definitely supported. Here's a diamond dependency modified from the tests in [test/town/lilac/flex_test.cljc](https://github.com/lilactown/flex/blob/99de18dacb460ea7dfa33f44bbe710a375cd30f6/test/town/lilac/flex_test.cljc)

```clojure
(def A (f/source 2))
(def B (f/signal (* @A @A)))
(def C (f/signal (+ @A 2)))
(def D (f/signal (* @C @C)))
(def fx (f/effect [] (prn (+ @B @D))))
```

In this diamond dependency, the propagation algorithm guarantees that each update to `A` will recalculate each signal only once.

## 3. signal wrapping everything

You can define your own macro to help you with this. :)

```clojure
(defmacro defs
  [sym body]
  `(def ~sym (flex/signal ~@body)))
```

## 4. No quick and easy way to subscribe to map values

> From the API it looks like in flex I need to create a signal for each an every key @_@ so that when a part of the map gets updated I get the appropriate update (that can propagate to all the signals). Otherwise my whole DAG gets recomputed on each map update (well I assume it'll terminate when identical values are returned). Ergonomically this is a bit of a mess. Basically if you have a .. simple? (finite/non-lazy) built in datatype as a source, you can be sure you'll automatically want signals for every single elements

I'm not sure that's a fair assumption, but it's easy to build a helper function for yourself if you like

```clojure
(defn sub
  [s f]
  (flex/signal (f s)))

(def params
  "The state atom"
  (flex/source
    {:input-dir   "./data/gauge/"
     :id-2-name   {1645240 5
                   8373316 3
                   7598252 4
                   1683136 2
                   8371400 1
                   1641660 6
                   7598984 7}
     :inside-d18o  "irregular-cave-samples.csv"
     :outside-d18o "manual-rain-samples.csv"
     :plot-height 400
     :plot-width 1000
     :blah        "blah"}))

(sub params :blah) ;; returns a signal that will track the value of `:blah` in the map
```

**Comment by kxygk** [comment]
This is fantastic! Thank you for getting back so fast and helping me iron out my misunderstandings

1. Oh it has to be connected! 
Okay! I had missed this point entirely :)) I sort of understand in the abstract the utility of having zero cost "lazy" signals - though I'm a bit hard pressed to think of a case I'd actually really need that (though this is likely just a lack of imagination on my part). Maybe they could be auto-connected by default? In either case, when I print in the REPL it also feels that that would implicitly create a "connection" to the REPL/io. Are you envisioning wanting to inspect a disposed/disconnected value for some reason? Again it doesn't seem like the hot path, and more of a cornercase. So seems maybe a strange default. I actually don't really how you develop interactively - it's possible we just have very different workflows. I like to poke at variable and function all over. Having to rig dummy connections each time would add a lot of friction (but something that could be easily automated with a bit of sugar/function/macros)

2. This was all do to my misunderstanding
After Q1 this whole part was just me being dumb - but thank you for spelling it out. And I will take a look at the tests as well

3. Thank you for the macro :)) I guess I could add extra things in here. Do you have any suggestions for how to make a hidden equivalent to the the dummy connection you made - but in a way that would not get garbage collected immediately when out of scope? I could see maybe stuffing it into the var meta data... but I'm not super familiar with writing macros or futzing with metadata tbh. Maybe you have better idea for how to create auto-connected signals

4. I don't think this accomplishes the desired effect. The issue is that say for instance you call `(sub params :blah)` in two different "higher order" signals -  each `sub` call would generate a new distinct signals. Ideally you'd want one signal for each key that is shared across all "consumers" of your atom/`source`. I guess the final behavior would be identical, but you're just making extra overhead (only an issue if the values/cells are large)

Thank you again for helping me work through the speed bumps. I can definitely start my rewrite with this. I really like your pragmatic/flexible approach

**Comment by lilactown** [comment]
> And please don't take it wrong. I don't mean to come and just say "Your defaults are wrong and you should do things different".

I would not take your posts this way! I assume that we are both learning. You, about my weird library, and me, about your weird use cases 😛 and each of us about our weird workflows.

I'll respond to a few key points in your last post

> Maybe they could be auto-connected by default? In either case, when I print in the REPL it also feels that that would implicitly create a "connection" to the REPL/io.

The downside of auto connecting by default is that you then do not have a way to detect when you should dispose of them and let them be garbage collected.

Take a simple example

```clojure
(def counter (flex/source 0))
(def counter-sq (flex/signal (* @counter @counter)))
```

When `counter-sq` is connected, the `counter` source needs to hold a reference to it so that when the source updates, it knows to go tell the `counter-sq` signal to recalculate its value. If we were to redefine `counter-sq`, that reference would remain lying around, meaning that as you develop over `counter-sq` signal over time you would accumulate many signals that would all be reacting to `counter` and never be garbage collected.

You might also end up having a program with 10s or hundreds of signals defined, and (IMO) they shouldn't all be reacting unless they're actually being used to produce some output, otherwise it's a waste. You can also have programs that dynamically create signals in reaction to other signals, which could end up with thousands of signals being created over time, and I want those to be able to be as close 0-cost as possible.

> I actually don't really how you develop interactively - it's possible we just have very different workflows. I like to poke at variable and function all over. Having to rig dummy connections each time would add a lot of friction (but something that could be easily automated with a bit of sugar/function/macros)

I think my use cases so far have been to create an app using signals and inspecting them once they're hooked up. So for instance I'll create a web UI that has some sources and signals and inspect them as they change over time. I think that I have tended to do less focused development on a specific signal, as most of my complex logic I put elsewhere.

I see the benefit of being able to test the logic of a signal without it being connected by derefing it outside of the context of a signal or effect. I think one of the things I wanted to communicate to the user is whether a signal in fact _is connected_ or not. Perhaps a better developer experience would be to do something like

```clojure
@counter-sq ;; => 2
;; prints out "WARNING: Signal dereferenced outside of reactive context"
```

What do you think?

> Do you have any suggestions for how to make a hidden equivalent to the the dummy connection you made - but in a way that would not get garbage collected immediately when out of scope?

I think the risk to creating a dummy connection is actually that it won't be garbage collected once out of scope, because signals and effects keep references to both their dependents and dependencies, so as long as the signal is in scope it will keep the effect in memory. The problem is that if you don't yourself store a reference to the effect, you can't dispose it and disconnect everything.

It sounds like your need isn't necessarily to have a connection that will auto update, but to be able to test the logic of your signals at the REPL. Is that right? If so, I could think about how to create a "non-reactive" branch that will calculate everything but print a warning and not cache its result without a connection, like I suggested above.

> Ideally you'd want one signal for each key that is shared across all "consumers" of your atom/source. (again I'm guessing this could be shoved into the var metadata somehow) I guess the final behavior would be identical, but you're just making extra overhead (only an issue if the values/cells are large)

Creating lots of little signals might not be a big deal, but you can also use `town.lilac.flex.memo/memoize` to create a cached `sub` function like you're saying.

```clojure
(require '[town.lilac.flex.memo :as flex.memo])
(def sub
  (flex.memo/memoize
    (fn [s f]
      (flex/signal (f @s))))

(def src (flex/source {:foo "bar" :baz 42}))
(def foo (sub src :foo))

(= foo (sub src :foo)) ;; => true
```
